Automate the Flask release process

Gudahtt opened the pull request:
A Flask release will now be published alongside each main extension release. The version of each Flask release will be the same as the extension version except it will have the suffix `-flask.0`.

The `sentry-publish.js` script was updated to support publishing non-main builds. It now takes additional `build-type` and `build-version` parameters, though only the former is used right now (the `build-version` parameter will be used for beta). It was converted to use `yargs` so that adding the new parameters would be easier.

Rather than create a separate release page for Flask, the Flask build is included alongside the main build in the main release page. We can integrate the changelog entries as well, prefixing any Flask-specific changes with `[Flask]`. This seemed easier than maintaining separate changelog, or doing any additional complex changelog parsing.

The Flask release will still get its own separate tag though. Just so it's totally clear which commit the release was created from.

Manual testing steps:
This is quite difficult to test manually, and I don't expect reviewers to do this. I can document my manual testing though.

I tested this by creating a new repository and going through the release process. The repository is here: https://github.com/Gudahtt/test-metamask-release

Here is the final successful workflow: https://app.circleci.com/pipelines/github/Gudahtt/test-metamask-release/36/workflows/07222244-3d62-491d-a967-6f7d0988678f
Here is the release page: https://github.com/Gudahtt/test-metamask-release/releases/tag/v11.0.5 (it's supposed to have the changelog as the body, but that's currently broken. Long-standing unrelated problem. We've been manually fixing it for each release.)
Here is the Flask tag: https://github.com/Gudahtt/test-metamask-release/releases/tag/v11.0.5-flask.0

Closes #13427

Gudahtt committed:
Automate the Flask release

A Flask release will now be published alongside each main extension
release. The version of each Flask release will be the same as the
extension version except it will have the suffix `-flask.0`.

Gudahtt commented:
This depends upon #13895

Gudahtt commented:
Leaving this in draft for now because I have not tested it yet. I will test this with a fork first.

rekmarks commented on .circleci/scripts/release-create-gh-release.sh:
```diff
@@ -26,25 +26,36 @@ function install_github_cli ()
+  local flask_filename
+  flask_filename="$(find ./builds-flask -type f -name 'metamask-flask-chrome-*.zip' -exec basename {} .zip \;)"
+
+  # Use substring parameter expansion to remove the first 22 characters ("metamask-extension-flask-")
+  echo "${flask_filename:22}"
```
Perhaps it would be better to type out the prefix we're trying to get the length of and then programmatically get its length?

Gudahtt replied:
Done: 065fc76ed97e8d08f378496f2d67dfdf08e0137d

rekmarks commented on development/sentry-upload-artifacts.sh:
```diff
@@ -51,6 +54,16 @@ function main {
         ;;
+      --dist-directory)
+        if [[ -z $2 ]]
+        then
+          printf "'dist-directory' option requires an argument.\\n" >&2
+          printf '%s\n' "${__SEE_HELP_MESSAGE__}" >&2
+          exit 1
+        fi
+        dist_directory="${2}"
```
Should we bother assigning `dist_directory` to `"dist"` if we require it anyway?

Gudahtt replied:
Sorry I don't understand what you mean

Gudahtt replied:
Oh, if you mean to suggest that it's a required parameter... it is not, it's optional. `dist` is the detault.

rekmarks reviewed (changes requested):
It's wonderful to see this! Found a couple issues, elaborated in comments.

Gudahtt commented on development/sentry-upload-artifacts.sh:
```diff
@@ -51,6 +54,16 @@ function main {
         shift
         ;;
+      --dist-directory)
+        if [[ -z $2 ]]
```
Yes, the `shift` call unsets it. `$1` is always "the current flag", and `$2` is always "the flag's value" (if it's set).

Gudahtt committed:
Programmatically remove build prefix

The create GH release Bash script derives the Flask version from the
Flask build filename by removing the build prefix, leaving just the
version. Rather than hard-coding the prefix size to remove, it is now
calculated programmatically so that it is easier to read and update.

Gudahtt committed:
Fix tag publishing

The tab publishing step used the wrong credentials, and didn't properly
identify the commit author. This has now been fixed.

rekmarks reviewed (approved):
LGTM!

mcmire commented on development/metamaskbot-build-announce.js:
```diff
@@ -3,7 +3,7 @@ const { promises: fs } = require('fs');
 const fetch = require('node-fetch');
 const glob = require('fast-glob');
-const VERSION = require('../dist/chrome/manifest.json').version; // eslint-disable-line import/no-unresolved
+const VERSION = require('../package.json').version;
```
Nice, good catch 👍🏻

mcmire reviewed (approved):
LGTM!
